### PR TITLE
Adding Root-based algorithm/tests/docs 

### PR DESCRIPTION
**Pull request by LBenzahia**
- [x] Enhance the light stemming Algorithm.
- [x] Add Root-based stemming Algorithm.
- [x]  Add docs.
- [x] Add tests.
- [ ] fix the problem of dists libs(python/java/jsx), Do not extract the roots.

### CODE REVIEW

File: algorithm/based_root_stemmer.sbl [hunk 1]
```diff
@@ -0,0 +1,906 @@
+/*
```
Comment (LBenzahia) [inline algorithm/based_root_stemmer.sbl]: @assem-ch   not ready

**Comment by assem-ch** [comment]
@LBenzahia pull from the upstream and merge fixing conflicts.  and push again

File: Makefile [hunk 2]
```diff
@@ -1,21 +1,25 @@
 SNOWBALL="./snowball/"
 ARABIC_STEMMER="algorithm/stemmer.sbl"
+ARABIC_BASED_ROOT_STEMMER = "algorithm/based_root_stemmer.sbl"
```
Comment (assem-ch) [inline Makefile]: ARABIC_ROOT_BASED_STEMMER , if you are using pycharm, use refactor -> rename feature

File: Makefile [hunk 3]
```diff
@@ -1,21 +1,25 @@
...
 VOCFILE="test_data/voc.txt"
 OUTPUTFILE="test_data/output.txt"
 GROUPINGFILE="test_data/grouping.txt"
-
+WORDS="golden_corpus/core/words.txt"
```
Comment (assem-ch) [inline Makefile]: CORPUS_WORDS

File: Makefile [hunk 4]
```diff
@@ -1,21 +1,25 @@
...
+STEMS="test_data/test_golden_corpus/stems_output.txt"
```
Comment (assem-ch) [inline Makefile]: Create a path for corpus , than use it to define CORPUS_WORDS, CORPUS_STEMS,...

File: Makefile [hunk 5]
```diff
@@ -24,17 +28,43 @@ download_data:
...
+
+build_based_root_stemmer:
+	@echo "Copying the algorithm of based-root stemmer to snowball..."
+	@cp $(ARABIC_BASED_ROOT_STEMMER) $(SNOWBALL)"algorithms/arabic/stem_Unicode.sbl"
+	@echo "Building based root stemmer ..."
 	@cd $(SNOWBALL); make
 
-run: build
+
+run_light_stemmer: build
```
Comment (assem-ch) [inline Makefile]: Keep this as `run`

File: Makefile [hunk 6]
```diff
@@ -24,17 +28,43 @@ download_data:
...
 	@echo "Put your words here:"
-	@cd snowball; ./stemwords -l ar 
+	@cd snowball; ./stemwords -l ar
+
+run_based_root_stemmer: build_based_root_stemmer
```
Comment (assem-ch) [inline Makefile]: this as `run_root`

File: algorithm/test/test_stemmer.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+r = open(ROOTS)
+w = open(WORDS)
+
+line1 = s.readline()
+line2 = w.readline()
+line3 = r.readline()
+cpt_stems = 0
+cpt_roots = 0
+i = 0
+while( line1 and line2 and line3 and i <= len(golden_corpus)- 1):
```
Comment (assem-ch) [inline algorithm/test/test_stemmer.py]: `i < len(golden_corpus)- 2` is clearer

File: algorithm/test/test_stemmer.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    golden_corpus = json.load(data_file)
+
+s = open(STEMS)
+r = open(ROOTS)
+w = open(WORDS)
+
+line1 = s.readline()
+line2 = w.readline()
+line3 = r.readline()
+cpt_stems = 0
```
Comment (assem-ch) [inline algorithm/test/test_stemmer.py]: cpt_stems = cpt_roots = 0

File: algorithm/test/test_stemmer.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+import os
+import sys
+import json
+import io
+
+# Make it work for Python 2+3 and with Unicode
```
Comment (assem-ch) [inline algorithm/test/test_stemmer.py]: good one 👍

**Comment by assem-ch** [comment]
Add a table in the readme file of 10 words of the same group and give its result using the stemmer and the rooter:
فأطفالكم اطفال طفل

File: docs/docs/index.md [hunk 10]
```diff
@@ -1,29 +1,155 @@
...
+  والطفلتين | طفل |طفل
+طفلتان | طفل | طفل
+
+## Contribute
+If you have a bug or suggestion, please consider to submit a report [here](https://github.com/assem-ch/arabicstemmer/issues/new). Open issues are listed [here](https://github.com/assem-ch/arabicstemmer/issues).
+
+## Authors
+* Assem Chelli `assem.ch[at]gmail[dot]com` Founder .
+
+## Contributers
```
Comment (assem-ch) [inline docs/docs/index.md]: remove me from contributors

Comment (LBenzahia) [inline docs/docs/index.md]: done !

**Comment by assem-ch** [comment]
Is it ready now?

what the new changes on the light_stemmer?

**Comment by LBenzahia** [comment]
Not yet for based-root algorithm.
For light_stemmer : 
1- Deals with جمع المؤنث السالم/مثنى المؤنث في حالة النصب وحالة الخفض as nouns, try this example: `الطالبات الطالبون الطالبتين الطالبتان الطالب الطالبة الطالبان الطالبين طالبتين` : 
the old version gives us : `طالب طالبت`. 
the new version gives one stem for this family : `طالب`.

**Comment by assem-ch** [comment]
whats not ready about the root-based stemmer

**Pull request merged**